SLING-9212 Distribution code checks for jcr:removeNode permissions on importer side for DELETE request

mohiaror opened the pull request:
Adding support for different multivalue properties for ADD and DELETE requests.

mohiaror committed:
SLING-9212 Distribution code checks for jcr:removeNode permissions on importer side for DELETE request

Adding support for different multivalue properties for ADD and DELETE requests.

actinium15 reviewed (changes requested):
hi @mohiaror,

I don’t think we want to make `jcrAddPrivileges` and `jcrDeletePrivileges` mandatory since it  defeats the purpose of having them from backwards compatibility perspective.

Either in the `PrivilegeDistributionRequestAuthorizationStrategy`, or in `PrivilegeDistributionRequestAuthorizationStrategyFactory`, it should be ensured that if the configs are missing, then existing behaviour should be picked up.

Also, please consider renaming the property-names/variables. These are _additional_ privileges and their names should reflect that - so maybe `additionalJcrPrivilegesForAdd` and `additionalJcrPrivilegesForDelete`?

mohiaror committed:
SLING-9212 Distribution code checks for jcr:removeNode permissions on importer side for DELETE request

Incorporating review feedback and not making additional privileges for ADD and DELETE requests mandatory.
Also adding test cases for added functionality.

mohiaror committed:
SLING-9212 Distribution code checks for jcr:removeNode permissions on importer side for DELETE request

Adding a null check.

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategy.java:
```diff
@@ -105,5 +113,14 @@ private String getClosestParent(Session session, String path) throws RepositoryE
         return null;
     }
 
+    private Privilege[] computePrivileges(AccessControlManager acMgr, String[] jcrPrivileges)
```
this method should be made `static` and `jcrPrivilege` should be supplied to it as its 2nd argument.

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyFactory.java:
```diff
@@ -59,13 +59,26 @@
+    /**
+     * privilege ADD request authorization strategy jcr privilege property
+     */
+    @Property(cardinality = 100, label = "Jcr Privilege Add", description = "Additional Jcr privileges to check for authorizing ADD distribution requests. The privilege is checked for the calling user session.")
```
```suggestion
    @Property(cardinality = 100, label = "Additional Jcr Privileges for Add", description = "Additional Jcr privileges to check for authorizing ADD distribution requests. The privilege is checked for the calling user session.")
```

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyFactory.java:
```diff
@@ -59,13 +59,26 @@
+     * privilege ADD request authorization strategy jcr privilege property
+     */
+    @Property(cardinality = 100, label = "Jcr Privilege Add", description = "Additional Jcr privileges to check for authorizing ADD distribution requests. The privilege is checked for the calling user session.")
+    private static final String JCR_ADD_PRIVILEGE = "additionalJcrPrivilegesForAdd";
```
```suggestion
    private static final String JCR_ADD_PRIVILEGES = "additionalJcrPrivilegesForAdd";
```

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyFactory.java:
```diff
@@ -59,13 +59,26 @@
+    /**
+     * privilege DELETE request authorization strategy jcr privilege property
+     */
+    @Property(cardinality = 100, label = "Jcr Privilege Delete", description = "Additional Jcr privileges to check for authorizing DELETE distribution requests. The privilege is checked for the calling user session.")
```
```suggestion
    @Property(cardinality = 100, label = "Additional Jcr Privileges for Delete", description = "Additional Jcr privileges to check for authorizing DELETE distribution requests. The privilege is checked for the calling user session.")
```

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyFactory.java:
```diff
@@ -59,13 +59,26 @@
+     * privilege DELETE request authorization strategy jcr privilege property
+     */
+    @Property(cardinality = 100, label = "Jcr Privilege Delete", description = "Additional Jcr privileges to check for authorizing DELETE distribution requests. The privilege is checked for the calling user session.")
+    private static final String JCR_DELETE_PRIVILEGE = "additionalJcrPrivilegesForDelete";
```
```suggestion
    private static final String JCR_DELETE_PRIVILEGES = "additionalJcrPrivilegesForDelete";
```

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyFactory.java:
```diff
@@ -59,13 +59,26 @@
     public void activate(BundleContext context, Map<String, Object> config) {
         String jcrPrivilege = PropertiesUtil.toString(config.get(JCR_PRIVILEGE), null);
-        authorizationStrategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        String[] jcrAddPrivilege = PropertiesUtil.toStringArray(config.get(JCR_ADD_PRIVILEGE), null);
```
```suggestion
        String[] jcrAddPrivileges = PropertiesUtil.toStringArray(config.get(JCR_ADD_PRIVILEGES), null);
```

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyFactory.java:
```diff
@@ -59,13 +59,26 @@
         String jcrPrivilege = PropertiesUtil.toString(config.get(JCR_PRIVILEGE), null);
-        authorizationStrategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        String[] jcrAddPrivilege = PropertiesUtil.toStringArray(config.get(JCR_ADD_PRIVILEGE), null);
+        String[] jcrDeletePrivilege = PropertiesUtil.toStringArray(config.get(JCR_DELETE_PRIVILEGE), null);
```
```suggestion
        String[] jcrDeletePrivileges = PropertiesUtil.toStringArray(config.get(JCR_DELETE_PRIVILEGES), null);
```

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyFactory.java:
```diff
@@ -59,13 +59,26 @@
-        authorizationStrategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        String[] jcrAddPrivilege = PropertiesUtil.toStringArray(config.get(JCR_ADD_PRIVILEGE), null);
+        String[] jcrDeletePrivilege = PropertiesUtil.toStringArray(config.get(JCR_DELETE_PRIVILEGE), null);
+        authorizationStrategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, jcrAddPrivilege, jcrDeletePrivilege);
```
```suggestion
        authorizationStrategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, jcrAddPrivileges, jcrDeletePrivileges);
```

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -97,7 +97,34 @@ public void testPermissionOnAdd() throws Exception {
         String[] paths = new String[]{"/foo"};
         for (String path : paths) {
-            when(acm.hasPrivileges(path, new Privilege[]{privilege, jcrReadPrivilege})).thenReturn(true);
+            when(acm.hasPrivileges(path, new Privilege[]{jcrReadPrivilege, privilege})).thenReturn(true);
```
Unchanged test-setup and asserts are crucial in clarifying that the changes are backwards compatible.
Though I _know_ that's not an issue here, still, rather than changing the test, please swap the order in which the privileges are added in `PrivilegeDistributionRequestAuthorizationStrategy.computePrivileges` itself and revert the changes here.

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -36,19 +36,21 @@
 public class PrivilegeDistributionRequestAuthorizationStrategyTest {
 
+    String jcrPrivilege = "foo";
+    String[] additionalJcrPrivilegesForAdd;
```
consider removing fields `jcrPrivilege`, `additionalJcrPrivilegesForAdd` and `additionalJcrPrivilegesForDelete`. Restore earlier, method-local `jcrPrivilege` and pass `null` directly while intiallizing `PrivilegeDistributionRequestAuthorizationStrategy` in the existing tests asserting existing behaviour (of adding default JCR_READ/JCR_REMOVE_NODE privileges as applicable)

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -36,19 +36,21 @@
     public void testCheckPermissionWithoutSession() throws Exception {
-        String jcrPrivilege = "foo";
-        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, additionalJcrPrivilegesForAdd, additionalJcrPrivilegesForDelete);
```
```suggestion
        String jcrPrivilege = "foo";
        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, null, null);
```

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -36,19 +36,21 @@
     public void testCheckPermissionWithSession() throws Exception {
-        String jcrPrivilege = "foo";
-        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, additionalJcrPrivilegesForAdd, additionalJcrPrivilegesForDelete);
```
```suggestion
        String jcrPrivilege = "foo";
        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, null, null);
```

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -58,8 +60,7 @@ public void testCheckPermissionWithSession() throws Exception {
     public void testNoPermissionOnAdd() throws Exception {
-        String jcrPrivilege = "somePermission";
-        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, additionalJcrPrivilegesForAdd, additionalJcrPrivilegesForDelete);
```
```suggestion
        String jcrPrivilege = "somePermission";
        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, null, null);
```

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -82,8 +83,7 @@ public void testNoPermissionOnAdd() throws Exception {
     public void testPermissionOnAdd() throws Exception {
-        String jcrPrivilege = "somePermission";
-        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, additionalJcrPrivilegesForAdd, additionalJcrPrivilegesForDelete);
```
```suggestion
        String jcrPrivilege = "somePermission";
        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, null, null);
```

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -97,7 +97,34 @@ public void testPermissionOnAdd() throws Exception {
+    
+    @Test
+    public void testAdditionalPermissionsOnAdd() throws Exception {
+        additionalJcrPrivilegesForAdd = new String[] {"addPermission"};
```
```suggestion
        String jcrPrivilege = "somePermission";
        additionalJcrPrivilegesForAdd = new String[] {"addPermission"};
```

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -97,7 +97,34 @@ public void testPermissionOnAdd() throws Exception {
+    @Test
+    public void testAdditionalPermissionsOnAdd() throws Exception {
+        additionalJcrPrivilegesForAdd = new String[] {"addPermission"};
+        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, additionalJcrPrivilegesForAdd, additionalJcrPrivilegesForDelete);
```
```suggestion
        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, additionalJcrPrivilegesForAdd, null);
```

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -97,7 +97,34 @@ public void testPermissionOnAdd() throws Exception {
+        Privilege privilege = mock(Privilege.class);
+        when(acm.privilegeFromName(jcrPrivilege)).thenReturn(privilege);
+        Privilege additionalPrivilegeForAdd = mock(Privilege.class);
+        for (String privilegeAdd : additionalJcrPrivilegesForAdd) {
```
either this `for` loop should be dropped, or the test adjusted to actually create correct fixtures. Right now, if one were to add another permission to `additionalJcrPrivilegesForAdd` array, the fixture below will return _identical_ mock for both of them. (not to mention the test will likely error-out because the fixture for `acm.hasPrivileges` won't work due to unexpected arguments).
Similarly, if you decide that `for` loop _IS NOT_ to be dropped, use `Answer` to verify the arguments sent while setting up the fixture for `acm.hasPrivileges`

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -107,8 +134,7 @@ public void testPermissionOnAdd() throws Exception {
     public void testNoPermissionOnDelete() throws Exception {
-        String jcrPrivilege = "somePermission";
-        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, additionalJcrPrivilegesForAdd, additionalJcrPrivilegesForDelete);
```
```suggestion
        String jcrPrivilege = "somePermission";
        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, null, null);
```

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -132,22 +158,49 @@ public void testNoPermissionOnDelete() throws Exception {
     public void testPermissionOnDelete() throws Exception {
-        String jcrPrivilege = "somePermission";
-        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, additionalJcrPrivilegesForAdd, additionalJcrPrivilegesForDelete);
```
```suggestion
        String jcrPrivilege = "somePermission";
        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege, null, null);
```

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -132,22 +158,49 @@ public void testNoPermissionOnDelete() throws Exception {
+        Privilege privilege = mock(Privilege.class);
+        when(acm.privilegeFromName(jcrPrivilege)).thenReturn(privilege);
+        Privilege additionalPrivilegeForDelete = mock(Privilege.class);
+        for (String privilegeDelete : additionalJcrPrivilegesForDelete) {
```
similar arguments and suggestions as made for `testAdditionalPermissionsOnAdd`

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -132,22 +158,49 @@ public void testNoPermissionOnDelete() throws Exception {
+    
+    @Test
+    public void testAdditionalPermissionsOnDelete() throws Exception {
+        additionalJcrPrivilegesForDelete = new String[] {"deletePermission"};
```
```suggestion
        String jcrPrivilege = "somePermission";
        additionalJcrPrivilegesForDelete = new String[] {"deletePermission"};
```

mohiaror committed:
SLING-9212 Distribution code checks for jcr:removeNode permissions on importer side for DELETE request

Incorporating review feedback.

tmaret commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategy.java:
```diff
@@ -105,5 +113,14 @@ private String getClosestParent(Session session, String path) throws RepositoryE
         return null;
     }
 
+    private static Privilege[] computePrivileges(AccessControlManager acMgr, String[] jcrPrivileges, String jcrPrivilege)
```
Privileges are currently re-computed for every permission check. How about computing them once in the constructor ? You may also keep track of privileges in a `Set` to avoid potential duplicates.

actinium15 replied:
I think the primary reason they are recomputed today is because `String`=>`Privilege` needs a JCR session (`AccessControlManager`, which is derieved from `Session`) which isn't available in `PrivilegeDistributionRequestAuthorizationStrategy()` (or `PrivilegeDistributionRequestAuthorizationStrategyFactory.activate`).

IMHO using a service session for doing that might be an overkill.

tmaret replied:
Ah, good point. We could keep the string representation in a Set.

tmaret commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategy.java:
```diff
@@ -68,8 +76,8 @@ public void checkPermission(@NotNull ResourceResolver resourceResolver, @NotNull
         AccessControlManager acMgr = session.getAccessControlManager();
-
-        Privilege[] privileges = new Privilege[]{acMgr.privilegeFromName(jcrPrivilege), acMgr.privilegeFromName(Privilege.JCR_READ)};
+        additionalJcrPrivilegesForAdd = additionalJcrPrivilegesForAdd != null ? additionalJcrPrivilegesForAdd : new String[] {Privilege.JCR_READ};
```
The backward compatible JRC_READ default should be a default value of the `JCR_ADD_PRIVILEGES` property. 

```
value = {"jcr:read"}
```

mohiaror replied:
If the config does not contain values of additionalJcrPrivilegesForAdd or additionalJcrPrivilegesForDelete (as is the case with default config right now) then we switch to using the earlier defaults, viz., JCR_READ and JCR_REMOVE_NODE, respectively as captured in the code. Do you mean something else?

tmaret replied:
Instead of the switch in the code, I suggest to leverage the default value of the property.

```
@Property(value="jcr:read", ...)
private static final String JCR_ADD_PRIVILEGES = "additionalJcrPrivilegesForAdd";
```

mohiaror replied:
@tmaret Thanks. I have made the changes.

tmaret commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategy.java:
```diff
@@ -81,8 +89,8 @@ private void checkPermissionForAdd(Session session, String[] paths)
         AccessControlManager acMgr = session.getAccessControlManager();
-
-        Privilege[] privileges = new Privilege[]{acMgr.privilegeFromName(jcrPrivilege), acMgr.privilegeFromName(Privilege.JCR_REMOVE_NODE)};
+        additionalJcrPrivilegesForDelete = additionalJcrPrivilegesForDelete != null ? additionalJcrPrivilegesForDelete : new String[] {Privilege.JCR_REMOVE_NODE};
```
The backward compatible JRC_REMOVE_NODE default should be a default value of the `JCR_DELETE_PRIVILEGES` property.

mohiaror replied:
Same comment as above.

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategy.java:
```diff
@@ -37,12 +39,18 @@
     private final String jcrPrivilege;
 
-    public PrivilegeDistributionRequestAuthorizationStrategy(String jcrPrivilege) {
+    private String[] additionalJcrPrivilegesForAdd;
```
```suggestion
    private final String[] additionalJcrPrivilegesForAdd;
```

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategy.java:
```diff
@@ -37,12 +39,18 @@
-    public PrivilegeDistributionRequestAuthorizationStrategy(String jcrPrivilege) {
+    private String[] additionalJcrPrivilegesForAdd;
+
+    private String[] additionalJcrPrivilegesForDelete;
```
```suggestion
    private final String[] additionalJcrPrivilegesForDelete;
```

mohiaror replied:
@actinium15 I change the additionalJcrPrivilegesForDelete/additionalJcrPrivilegesForAdd arrays based on their nullability so can't use final here. Since they are used just once in the class, I do not see a point in using a separate variable for keeping up with the default value of these arrays.

actinium15 replied:
> I change the additionalJcrPrivilegesForDelete/additionalJcrPrivilegesForAdd arrays based on their nullability so can't use final here.

Um, but I don't see any reason why you'd need to do that. You can, in fact, you should move the initializations in the CTOR itself. Something like this:
```
--- a/src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategy.java
+++ b/src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategy.java
@@ -39,9 +39,9 @@ public class PrivilegeDistributionRequestAuthorizationStrategy implements Distri

     private final String jcrPrivilege;

-    private String[] additionalJcrPrivilegesForAdd;
+    private final String[] additionalJcrPrivilegesForAdd;

-    private String[] additionalJcrPrivilegesForDelete;
+    private final String[] additionalJcrPrivilegesForDelete;

     public PrivilegeDistributionRequestAuthorizationStrategy(String jcrPrivilege, String[] additionalJcrPrivilegesForAdd, String[] additionalJcrPrivilegesForDelete) {
         if (jcrPrivilege == null) {
@@ -49,8 +49,12 @@ public class PrivilegeDistributionRequestAuthorizationStrategy implements Distri
         }

         this.jcrPrivilege = jcrPrivilege;
-        this.additionalJcrPrivilegesForAdd = additionalJcrPrivilegesForAdd;
-        this.additionalJcrPrivilegesForDelete = additionalJcrPrivilegesForDelete;
+        this.additionalJcrPrivilegesForAdd =
+                additionalJcrPrivilegesForAdd != null ?
+                        additionalJcrPrivilegesForAdd : new String[] {Privilege.JCR_READ};;
+        this.additionalJcrPrivilegesForDelete =
+                additionalJcrPrivilegesForDelete != null ?
+                        additionalJcrPrivilegesForDelete : new String[] {Privilege.JCR_REMOVE_NODE};;
     }

     public void checkPermission(@NotNull ResourceResolver resourceResolver, @NotNull DistributionRequest distributionRequest) throws DistributionException {
@@ -76,7 +80,6 @@ public class PrivilegeDistributionRequestAuthorizationStrategy implements Distri
     private void checkPermissionForAdd(Session session, String[] paths)
             throws RepositoryException, DistributionException {
         AccessControlManager acMgr = session.getAccessControlManager();
-        additionalJcrPrivilegesForAdd = additionalJcrPrivilegesForAdd != null ? additionalJcrPrivilegesForAdd : new String[] {Privilege.JCR_READ};
         Privilege[] privileges = computePrivileges(acMgr, jcrPrivilege, additionalJcrPrivilegesForAdd);
         for (String path : paths) {
             if (!acMgr.hasPrivileges(path, privileges)) {
@@ -89,7 +92,6 @@ public class PrivilegeDistributionRequestAuthorizationStrategy implements Distri
     private void checkPermissionForDelete(Session session, String[] paths)
             throws RepositoryException, DistributionException {
         AccessControlManager acMgr = session.getAccessControlManager();
-        additionalJcrPrivilegesForDelete = additionalJcrPrivilegesForDelete != null ? additionalJcrPrivilegesForDelete : new String[] {Privilege.JCR_REMOVE_NODE};
         Privilege[] privileges = computePrivileges(acMgr, jcrPrivilege, additionalJcrPrivilegesForDelete);
         for (String path : paths) {

```

Would you think above is a bad idea? if yes, why? What am I missing here?

tmaret replied:
I think it's a good idea, inline with https://github.com/apache/sling-org-apache-sling-distribution-core/pull/40#discussion_r413626074 .

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategy.java:
```diff
@@ -105,5 +113,14 @@ private String getClosestParent(Session session, String path) throws RepositoryE
         return null;
     }
 
+    private static Privilege[] computePrivileges(AccessControlManager acMgr, String[] jcrPrivileges, String jcrPrivilege)
```
```suggestion
    private static Privilege[] computePrivileges(AccessControlManager acMgr, String jcrPrivilege, String[] additionalJcrPrivileges)
```

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategy.java:
```diff
@@ -105,5 +113,14 @@ private String getClosestParent(Session session, String path) throws RepositoryE
+            throws RepositoryException {
+        ArrayList<Privilege> privileges = new ArrayList<Privilege>();
+        privileges.add(acMgr.privilegeFromName(jcrPrivilege));
+        for (String privilege : jcrPrivileges) {
```
```suggestion
        for (String privilege : additionalJcrPrivileges) {
```

mohiaror committed:
SLING-9212 Distribution code checks for jcr:removeNode permissions on importer side for DELETE request

Incorporating review feedback.

mohiaror committed:
SLING-9212 Distribution code checks for jcr:removeNode permissions on importer side for DELETE request

Incorporating review comments.

mohiaror commented:
@actinium15 , @tmaret  I have incorporated the suggestions. Please check if the changes look good.

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyFactory.java:
```diff
@@ -59,13 +60,28 @@
     public void activate(BundleContext context, Map<String, Object> config) {
         String jcrPrivilege = PropertiesUtil.toString(config.get(JCR_PRIVILEGE), null);
-        authorizationStrategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        String[] jcrAddPrivileges = PropertiesUtil.toStringArray(config.get(JCR_ADD_PRIVILEGES), null);
```
Since the property now has a default, I'd suggest losing the `null` for clarity to the reader...
```suggestion
        String[] jcrAddPrivileges = PropertiesUtil.toStringArray(config.get(JCR_ADD_PRIVILEGES));
```

actinium15 commented on src/main/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyFactory.java:
```diff
@@ -59,13 +60,28 @@
         String jcrPrivilege = PropertiesUtil.toString(config.get(JCR_PRIVILEGE), null);
-        authorizationStrategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        String[] jcrAddPrivileges = PropertiesUtil.toStringArray(config.get(JCR_ADD_PRIVILEGES), null);
+        String[] jcrDeletePrivileges = PropertiesUtil.toStringArray(config.get(JCR_DELETE_PRIVILEGES), null);
```
Since the property now has a default, I'd suggest losing the `null` for clarity to the reader...
```suggestion
        String[] jcrDeletePrivileges = PropertiesUtil.toStringArray(config.get(JCR_DELETE_PRIVILEGES));
```

mohiaror committed:
SLING-9212 Distribution code checks for jcr:removeNode permissions on importer side for DELETE request

Incorporating review comment.

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -131,23 +175,53 @@ public void testNoPermissionOnDelete() throws Exception {
+    public void testDefaultPermissionOnDelete() throws Exception {
         String jcrPrivilege = "somePermission";
-        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        String DEFAULT_DELETE_PRIVELEGE = "jcr:removeNode";
```
Do we really need this? Can't we just use Privilege.JCR_REMOVE_NODE?
If we can't, please fix the typo (misspelt "privilege")

mohiaror replied:
Using Privilege.JCR_REMOVE_NODE will give it an expanded form which will not look good on the config manager UI and might confuse the admin that this is the way the privileges are supposed to be configured.

I will take care of the typo.

actinium15 commented on src/test/java/org/apache/sling/distribution/agent/impl/PrivilegeDistributionRequestAuthorizationStrategyTest.java:
```diff
@@ -81,17 +95,19 @@ public void testNoPermissionOnAdd() throws Exception {
+    public void testDefaultPermissionOnAdd() throws Exception {
         String jcrPrivilege = "somePermission";
-        PrivilegeDistributionRequestAuthorizationStrategy strategy = new PrivilegeDistributionRequestAuthorizationStrategy(jcrPrivilege);
+        String DEFAULT_READ_PRIVELEGE = "jcr:read";
```
Do we really need this? Can't we just use Privilege.JCR_READ?
If we can't, please fix the typo (misspelt "privilege")

actinium15 reviewed (approved):
Thanks for working on my suggestions! Left a minor comment...

mohiaror committed:
SLING-9212 Distribution code checks for jcr:removeNode permissions on importer side for DELETE request

Fixing typo.

tmaret commented:
Thanks @mohiaror and @actinium15 ! Merged.
